### ISSUE TITLE
Latest nightly report

### ISSUE THREAD

**Issue by amircodota** [rust-lang/rust-clippy#12069]
### Summary

```warning: this `let...else` may be rewritten with the `?` operator
   --> cloud2/snippet_suggestions/src/snippet_prompt/mod.rs:229:9
    |
229 | /         let Some(enricher_args): Option<EnricherArgs> = args.into() else {
230 | |             return None;
231 | |         };
    | |__________^ help: replace it with: `let enricher_args = args.into()?;`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#question_mark
    = note: `#[warn(clippy::question_mark)]` on by default```

This does not work because the `From` impl does not match

`args` is of type `args: &SnippetAutocompleteRequest,`

This is the `From` impl
```
impl<'a> From<&'a SnippetAutocompleteRequest<'a>> for Option<EnricherArgs<'a>> {
```

### Lint Name

clippy::question_mark

### Version

```text
cargo +nightly --version                                                                                                                          cargo 1.77.0-nightly (ac6bbb332 2023-12-26)
```

**Comment by PartiallyTyped** [comment]
Hi, if you don't mind, could you provide a minimal example?

**Comment by amircodota** [comment]
Hi

Here's a simple main.rs

```rust
fn main() {
    _ = my_func();
}

fn my_func() -> Option<MyStruct2> {
    let Some(my): Option<MyStruct2> = MyStruct.into() else { return None };

    my.do_something();

    Some(my)
}

struct MyStruct;
struct MyStruct2;

impl MyStruct2 {
    fn do_something(&self) {
        println!("do something");
    }

}

impl From<MyStruct> for Option<MyStruct2> {
    fn from(_: MyStruct) -> Self {
        Some(MyStruct2)
    }
}
```

If I run `cargo clippy` it warns me that

```
warning: this `let...else` may be rewritten with the `?` operator
 --> src/main.rs:6:5
  |
6 |     let Some(my): Option<MyStruct2> = MyStruct.into() else { return None };
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `let my = MyStruct.into()?;`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#question_mark
  = note: `#[warn(clippy::question_mark)]` on by default
```

However, if I change the line in question to 

```rust
let my: MyStruct2 = MyStruct.into()?;
```

it does not compile

```
error[E0283]: type annotations needed
 --> src/main.rs:6:34
  |
6 |     let my: MyStruct2 = MyStruct.into()?;
  |                                  ^^^^
  |
  = note: cannot satisfy `_: From<MyStruct>`
  = note: required for `MyStruct` to implement `Into<_>`
help: try using a fully qualified path to specify the expected types
  |
6 |     let my: MyStruct2 = <MyStruct as Into<T>>::into(MyStruct)?;
  |                         ++++++++++++++++++++++++++++        ~

For more information about this error, try `rustc --explain E0283`.
```

**NOTE:**
In this simple case, clippy warns me even in the latest stable (rust 1.75). However, in my (more complex) real code, clippy does not warn, however, clippy nightly does warn.
